examples/crypto-accelerator.p4

```diff
@@ -0,0 +1,508 @@
+/*
+Copyright 2022 Advanced Micro Devices, Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <core.p4>
+#include "../pna.p4"
+
+/// Crypto accelerator Extern
+enum bit<8> crypto_algorithm_e {
+    AES_GCM = 1
+}
+enum bit<8> crypto_results_e {
+    SUCCESS = 0,
+    AUTH_FAILURE = 1,
+    HW_ERROR = 2
+}
+
+// special offset value to indicate that ICV is after payload
+#define ICV_AFTER_PAYLOAD ((int<32>)-1)
+
+extern crypto_accelerator {
+    /// Some methods provided in this object may be specific to an algorithm used.
+    /// Compiler may be able to check and warn/error when incorrect methods are used
+
+    /// Constructor
+    crypto_accelerator(crypto_algorithm_e algo);
+
+    /// Security association index for the security session
+    /// Some implementations do not need it. In that case this method should result in no-op
+    void set_sa_index<T>(in T sa_index);
+
+    /// Set the initialization data based on protocol used.
+    /// E.g. salt, random number/ counter for ipsec
+    void set_iv<T>(in T iv);
+    void set_key<T,S>(in T key, in S key_size);
+
+    /// Authentication data format is protocol specific
+    /// Add this data to the packet and provide its offset and length using the
+    /// following methods
+    /// The format of the auth data is not specified/mandated by this object definition
+    void set_auth_data_offset<T>(in T offset);
+    void set_auth_data_len<T>(in T len);
+
+    /// Alternatively: Following method can be used to consturct protocol specific auth_data and
+    /// provide it to the engine.
+    void add_auth_data<T>(in T auth_data);
+
+    /// Auth trailer aka ICV is added by the engine after doing encryption operation
+    /// Specify icv location - when a wire protocol wants to add ICV in a specific location (e.g. AH)
+    /// The following methods can be used to specify the location of ICV in the packet
+    /// special offset ICV_AFTER_PAYLOAD indicates ICV is after the payload
+    void set_icv_offset<T>(in T offset);
+    void set_icv_len<T>(in T len);
+
+    /// Setup payload to be encrypted/decrypted
+    void set_payload_offset<T>(in T offset);
+    void set_payload_len<T>(in T len);
+
+    /// Operation
+    void encrypt<T>(in T enable_auth);
+    void decrypt<T>(in T enable_auth);
```

Comment:
Following today's discussion, perhaps these should explicitly say these operations are asynchronous, e.g. `mark_for_encrypt`, `encrypt_async`, `prime_encryption`, etc.

Reply:
All methods are async as the assumption as stated in the PR's description is that the crypto accelerator is at the end of the pipeline. Since all methods are async, I think it is not necessary to name them as async. This also seems to be the convention with other externs such as drop_packet() or send_to_port() These actions take effect after completing the execution of the current pipeline (including deparser).
See description of drop_packet here - 
https://github.com/p4lang/pna/blob/main/pna.p4#L642

I can add more description to this effect in the comments.. but prefer not to change the names.
When we add sync method we should name it as such since that does not fit current architecture.

Reply:
changed the names as encrypt -> enable_encrypt, decrypt -> enable_decrypt and also added some comments to indicate that this operation happens after deparser.